reasoner_pydantic/utils.py

```diff
@@ -0,0 +1,147 @@
+import collections
+from typing import Callable, Dict, List, Generic, TypeVar
+
+from pydantic import PrivateAttr
+from pydantic.generics import GenericModel
+
+KeyType = TypeVar('KeyType')
+ValueType = TypeVar('ValueType')
+
+class HashableMapping(
+        GenericModel,
+        Generic[KeyType, ValueType],
+        collections.abc.MutableMapping,
+):
+    """
+    Custom class that implements MutableMapping and is hashable
+
+    Hash will be recomputed if items are updated or deleted. The
+    hash can be considered valid if the values are immutable.
+    """
+    __root__: Dict[KeyType, ValueType]
+
+    _hash: int = PrivateAttr(default=None)
+    _invalidate_hook: Callable = PrivateAttr(default=None)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        for value in self.__root__.values():
+            if hasattr(value, "_invalidate_hook"):
+                value._invalidate_hook = self.invalidate_hash
+
+    def __getitem__(self, k):
+        return self.__root__[k]
+    def __iter__(self):
+        return iter(self.__root__)
+    def __len__(self):
+        return len(self.__root__)
+
+    def __setitem__(self, k, v):
+        self.invalidate_hash()
+        if hasattr(v, "_invalidate_hook"):
+            v._invalidate_hook = self.invalidate_hash
+        self.__root__[k] = v
+    def __delitem__(self, k):
+        self.invalidate_hash()
+        del self.__root__[k]
+
+    def __hash__(self):
+        if self._hash is None:
+            h = 0
+            for key, value in self.__root__.items():
+                h ^= hash((key, value))
```

Comment:
are you worried about getting `__root__.items()` in a different order here? the recommend way to make a hash of multiple values it to hash the tuple of those values.

Reply:
I'm a bit worried about extra memory usage from hashing using the tuple constructor instead of one at a time, but since premature optimization is the root of all evil I'll revert it and we can change it later if that's an issue.